WhitespaceAround on enhanced switch empty block

I have read check documentation: https://checkstyle.org/config_xxxxxx.html#NameOfAffectedCheck
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac YOUR_FILE.java
#[[MAKE SURE THERE IS SUCCESSFUL COMPILATION]]
```

/var/tmp $ cat config.xml
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <property name="haltOnException" value="true"/>
    <property name="severity" value="error"/>

    <module name="FileTabCharacter">
        <property name="eachLine" value="true"/>
    </module>
    <module name="NewlineAtEndOfFile"/>
    <module name="SuppressWarningsFilter"/>
    <module name="UniqueProperties"/>

    <module name="TreeWalker">
        <module name="SuppressWarningsHolder"/>

        <module name="WhitespaceAround">
            <property name="allowEmptyCatches" value="true"/>
            <property name="allowEmptyConstructors" value="true"/>
            <property name="allowEmptyMethods" value="true"/>
            <property name="allowEmptyLambdas" value="true"/>
            <property name="allowEmptyLoops" value="true"/>
            <property name="allowEmptyTypes" value="true"/>
            <property name="ignoreEnhancedForColon" value="false"/>
        </module>
    </module>
</module>
```

/var/tmp $ cat YOUR_FILE.java
```java
public class WhitespaceAround {
    private void switchTest() {
        switch ("foo") {
            case "A" -> {}
            case "B" -> System.out.println("B");
            default -> System.out.println("other");
        }
    }
}
```
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-X.XX-all.jar -c config.xml YOUR_FILE.java
```bash
Starting audit...
[ERROR] /private/var/tmp/WhitespaceAround.java:4:25: '{' is not followed by whitespace. [WhitespaceAround]
[ERROR] /private/var/tmp/WhitespaceAround.java:4:26: '}' is not preceded with whitespace. [WhitespaceAround]
Audit done.
Checkstyle ends with 2 errors.
```

For Windows users, please use `type` instead of `cat` and run

```cmd
set RUN_LOCALE="-Duser.language=en -Duser.country=US"
java %RUN_LOCALE% -jar checkstyle-X.XX-all.jar -c config.xml YOUR_FILE.java
```

in place of the last 2 commands above.

---
My expectation was that one of the allowEmpty* properties would stop the violation  on the line 
```java
case "A" -> {}
```

---

Still not clear ???
see example - https://checkstyle.org/report_issue.html#How_to_report_a_bug.3F

ATTENTION: FAILURE TO FOLLOW THE ABOVE TEMPLATE WILL RESULT IN THE ISSUE BEING CLOSED.

Violations for WhitespaceAround are shown for both types of switch statements:

```bash
➜  src cat MyClass.java
public class MyClass {
    private void switchTest1() {
        switch ("foo") {
            case "A" -> {}
            case "B" -> System.out.println("B");
            default -> System.out.println("other");
        }
    }

    private void switchTest2() {
        switch ("foo") {
            case "A" : {}
            case "B" : System.out.println("B");
            default : System.out.println("other");
        }
    }

    public static void main(String[] args) {
    }

}
➜  src java MyClass.java
➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://checkstyle.sourceforge.net/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <property name="haltOnException" value="true"/>
    <property name="severity" value="error"/>

    <module name="FileTabCharacter">
        <property name="eachLine" value="true"/>
    </module>
    <module name="NewlineAtEndOfFile"/>
    <module name="SuppressWarningsFilter"/>
    <module name="UniqueProperties"/>

    <module name="TreeWalker">
        <module name="SuppressWarningsHolder"/>

        <module name="WhitespaceAround">
            <property name="allowEmptyCatches" value="true"/>
            <property name="allowEmptyConstructors" value="true"/>
            <property name="allowEmptyMethods" value="true"/>
            <property name="allowEmptyLambdas" value="true"/>
            <property name="allowEmptyLoops" value="true"/>
            <property name="allowEmptyTypes" value="true"/>
            <property name="ignoreEnhancedForColon" value="false"/>
        </module>
    </module>
</module>

➜  src java -jar checkstyle-8.41-all.jar -c config.xml MyClass.java
Starting audit...
[ERROR] Tester/src/MyClass.java:4:25: '{' is not followed by whitespace. [WhitespaceAround]
[ERROR] Tester/src/MyClass.java:4:26: '}' is not preceded with whitespace. [WhitespaceAround]
[ERROR] Tester/src/MyClass.java:12:24: '{' is not followed by whitespace. [WhitespaceAround]
[ERROR] Tester/src/MyClass.java:12:25: '}' is not preceded with whitespace. [WhitespaceAround]
Audit done.
Checkstyle ends with 4 errors.

```
There is no `allow*` property for empty switch case blocks, so we would need to implement a new property in WhitespaceAround like `allowEmptySwitch`.

`LAMBDA` token in this case is really the arrow operator (definite assignment, so that `break` is not needed), not really a "lambda" as in a lambda expression, and should not be considered as such when considering whitespace around tokens. 

> Here the problem is that LAMBDA is not the parent of SLIST Token.

The `LAMBDA` token should not be the parent of  `SLIST` [they should be siblings.](https://docs.oracle.com/javase/specs/jls/se12/preview/switch-expressions.html)

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheck.java
@@ -47,7 +47,8 @@ import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
  * <p>
  * may optionally be exempted from the policy using the {@code allowEmptyMethods},
  * {@code allowEmptyConstructors}, {@code allowEmptyTypes}, {@code allowEmptyLoops},
- * {@code allowEmptyLambdas} and {@code allowEmptyCatches} properties.
+ * {@code allowEmptyLambdas}, {@code allowEmptyCatches}
+ * and {@code allowEmptySwitchBlockStatements} properties.
  * </p>
  *
  * <p>
@@ -103,6 +104,12 @@ import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
  * </li>
  * <li>
  * Property {@code allowEmptyMethods} - Allow empty method bodies.
+ * Type is {@code boolean}.
+ * Default value is {@code false}.
+ * </li>
+ * <li>
+ * Property {@code allowEmptySwitchBlockStatements} - Allow empty switch blocks
+ * and block statements.
  * Type is {@code boolean}.
  * Default value is {@code false}.
  * </li>
@@ -275,6 +282,8 @@ public class WhitespaceAroundCheck extends AbstractCheck {
     private boolean allowEmptyLambdas;
     /** Allow empty catch bodies. */
     private boolean allowEmptyCatches;
+    /** Allow empty switch blocks and block statements. */
+    private boolean allowEmptySwitchBlockStatements;
     /**
      * Ignore whitespace around colon in
      * <a href="https://docs.oracle.com/javase/specs/jls/se11/html/jls-14.html#jls-14.14.2">
@@ -480,6 +489,16 @@ public class WhitespaceAroundCheck extends AbstractCheck {
         allowEmptyCatches = allow;
     }
 
+    /**
+     * Setter to allow empty switch blocks and block statements.
+     *
+     * @param allow {@code true} to allow empty switch case and default blocks.
+     * @since 10.19.0
+     */
+    public void setAllowEmptySwitchBlockStatements(boolean allow) {
+        allowEmptySwitchBlockStatements = allow;
+    }
+
     @Override
     public void visitToken(DetailAST ast) {
         final int currentType = ast.getType();
@@ -606,7 +625,8 @@ public class WhitespaceAroundCheck extends AbstractCheck {
                 || isEmptyCtorBlockCheckedFromRcurly(ast)
                 || isEmptyLoop(ast, parentType)
                 || isEmptyLambda(ast, parentType)
-                || isEmptyCatch(ast, parentType);
+                || isEmptyCatch(ast, parentType)
+                || isEmptySwitchBlockStatement(ast);
     }
 
     /**
@@ -725,6 +745,41 @@ public class WhitespaceAroundCheck extends AbstractCheck {
      */
     private boolean isEmptyCatch(DetailAST ast, int parentType) {
         return allowEmptyCatches && isEmptyBlock(ast, parentType, TokenTypes.LITERAL_CATCH);
+    }
+
+    /**
+     * Tests if the given {@code DetailAst} is part of an allowed empty
+     * switch case or default block.
+     *
+     * @param ast the {@code DetailAst} to test.
+     * @return {@code true} if {@code ast} makes up part of an allowed
+     *         empty switch case or default block.
+     */
+    private boolean isEmptySwitchBlockStatement(DetailAST ast) {
+        final boolean isEmptySwitchBlockStatement;
+
+        if (allowEmptySwitchBlockStatements) {
+            final DetailAST parent = ast.getParent();
+            final DetailAST grandParent = parent.getParent();
+
+            final boolean isEmptyCaseInSwitchRule =
+                    isEmptyBlock(ast, parent.getType(), TokenTypes.SWITCH_RULE);
+
+            final boolean isEmptyCaseGroupCheckedFromLcurly =
+                    isEmptyBlock(ast, grandParent.getType(), TokenTypes.CASE_GROUP);
+
+            final boolean isEmptyCaseGroupCheckedFromRcurly =
+                    parent.getFirstChild().getType() == TokenTypes.RCURLY
+                      && grandParent.getParent().getType() == TokenTypes.CASE_GROUP;
+
+            isEmptySwitchBlockStatement = isEmptyCaseInSwitchRule
+                    || isEmptyCaseGroupCheckedFromLcurly || isEmptyCaseGroupCheckedFromRcurly;
+        }
+        else {
+            isEmptySwitchBlockStatement = false;
+        }
+
+        return isEmptySwitchBlockStatement;
     }
 
     /**

--- a/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/whitespace/WhitespaceAroundCheck.xml
+++ b/src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/whitespace/WhitespaceAroundCheck.xml
@@ -25,7 +25,8 @@
  &lt;p&gt;
  may optionally be exempted from the policy using the {@code allowEmptyMethods},
  {@code allowEmptyConstructors}, {@code allowEmptyTypes}, {@code allowEmptyLoops},
- {@code allowEmptyLambdas} and {@code allowEmptyCatches} properties.
+ {@code allowEmptyLambdas}, {@code allowEmptyCatches}
+ and {@code allowEmptySwitchBlockStatements} properties.
  &lt;/p&gt;
 
  &lt;p&gt;
@@ -73,6 +74,12 @@
             </property>
             <property default-value="false" name="allowEmptyMethods" type="boolean">
                <description>Allow empty method bodies.</description>
+            </property>
+            <property default-value="false"
+                      name="allowEmptySwitchBlockStatements"
+                      type="boolean">
+               <description>Allow empty switch blocks
+ and block statements.</description>
             </property>
             <property default-value="false" name="allowEmptyTypes" type="boolean">
                <description>Allow empty class, interface and enum bodies.</description>

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/WhitespaceAroundCheckTest.java
@@ -610,4 +610,74 @@ public class WhitespaceAroundCheckTest
         verifyWithInlineConfigParser(
                 getNonCompilablePath("InputWhitespaceAroundUnnamedPattern.java"), expected);
     }
+
+    @Test
+    public void testSwitchCasesParens() throws Exception {
+        final String[] expected = {
+            "33:21: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "33:22: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "37:22: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "37:23: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "47:23: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "47:24: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "51:24: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "51:25: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "59:23: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "59:48: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "68:22: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "68:47: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "76:9: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "76:10: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "81:9: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "81:34: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "89:21: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "89:22: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "89:24: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "89:25: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "96:21: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "96:23: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "100:21: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "100:22: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "100:22: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "{"),
+            "100:23: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "}"),
+            "100:23: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "100:24: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "108:23: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "108:24: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "115:13: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "}"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputWhitespaceAroundSwitchCasesParens.java"),
+                expected);
+    }
+
+    @Test
+    public void testSwitchCasesParensWithAllowEmptySwitchBlockStatements() throws Exception {
+        final String fileName =
+                "InputWhitespaceAroundSwitchCasesParensWithAllowEmptySwitchBlockStatements.java";
+
+        final String[] expected = {
+            "49:23: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "49:48: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "59:22: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "59:47: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "68:9: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "68:10: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "73:9: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "73:34: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "84:21: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "84:23: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "88:21: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "88:22: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "88:22: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "{"),
+            "88:23: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "}"),
+            "88:23: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "88:24: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+            "96:23: " + getCheckMessage(MSG_WS_NOT_FOLLOWED, "{"),
+            "96:24: " + getCheckMessage(MSG_WS_NOT_PRECEDED, "}"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(fileName),
+                expected);
+    }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundSwitchCasesParens.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundSwitchCasesParens.java
@@ -1,0 +1,118 @@
+/*
+WhitespaceAround
+allowEmptyConstructors = true
+allowEmptyMethods = (default)false
+allowEmptyTypes = (default)false
+allowEmptyLoops = (default)false
+allowEmptyLambdas = (default)false
+allowEmptyCatches = (default)false
+allowEmptySwitchBlockStatements = (default)false
+ignoreEnhancedForColon = (default)true
+tokens = (default)ASSIGN, BAND, BAND_ASSIGN, BOR, BOR_ASSIGN, BSR, BSR_ASSIGN, BXOR, \
+         BXOR_ASSIGN, COLON, DIV, DIV_ASSIGN, DO_WHILE, EQUAL, GE, GT, LAMBDA, LAND, \
+         LCURLY, LE, LITERAL_CATCH, LITERAL_DO, LITERAL_ELSE, LITERAL_FINALLY, \
+         LITERAL_FOR, LITERAL_IF, LITERAL_RETURN, LITERAL_SWITCH, LITERAL_SYNCHRONIZED, \
+         LITERAL_TRY, LITERAL_WHILE, LOR, LT, MINUS, MINUS_ASSIGN, MOD, MOD_ASSIGN, \
+         NOT_EQUAL, PLUS, PLUS_ASSIGN, QUESTION, RCURLY, SL, SLIST, SL_ASSIGN, SR, \
+         SR_ASSIGN, STAR, STAR_ASSIGN, LITERAL_ASSERT, TYPE_EXTENSION_AND
+
+
+*/
+
+//non-compiled with javac: Compilable with Java17
+package com.puppycrawl.tools.checkstyle.checks.whitespace.whitespacearound;
+
+public class InputWhitespaceAroundSwitchCasesParens {
+
+    void m(int k) {
+        switch (k) {
+            case 1 :
+                System.out.println("1");
+            case 2: { }
+            case 3:
+            case 4: {}
+            // 2 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            default: {}
+            // 2 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+        }
+    }
+    void m2(int k) {
+        switch (k) {
+            case 1 -> System.out.println("1");
+            case 2 -> { }
+            case 3 -> {}
+            // 2 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            default -> {}
+            // 2 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+        }
+    }
+    void m3(int k) {
+        switch (k) {
+            case 1 -> {System.out.println("1");}
+            // 2 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            default -> { }
+        }
+    }
+    void m4(int k) {
+        switch (k) {
+            case 1 : {System.out.println("1");}
+            // 2 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            default : { }
+        }
+    }
+    void m5(int k) {
+        {}
+        // 2 violations above:
+        //                  ''{' is not followed by whitespace.'
+        //                  ''}' is not preceded with whitespace.'
+
+        {System.out.println("1");}
+        // 2 violations above:
+        //                  ''{' is not followed by whitespace.'
+        //                  ''}' is not preceded with whitespace.'
+
+    }
+    void m6(int k) {
+        switch (k) {
+            case 1: {} {}
+            // 4 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            case 2: { } { }
+            case 3: {;}
+            // 2 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            case 4: {{}}
+            // 6 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''{' is not followed by whitespace.'
+            //                  ''{' is not preceded with whitespace.'
+            //                  ''}' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            case 5: { {} }
+            // 2 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            case 6: { { } }
+            case 7:  {
+
+            }break; // violation, ''}' is not followed by whitespace.'
+        }
+    }
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundSwitchCasesParensWithAllowEmptySwitchBlockStatements.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/whitespacearound/InputWhitespaceAroundSwitchCasesParensWithAllowEmptySwitchBlockStatements.java
@@ -1,0 +1,106 @@
+/*
+WhitespaceAround
+allowEmptyConstructors = true
+allowEmptyMethods = (default)false
+allowEmptyTypes = (default)false
+allowEmptyLoops = (default)false
+allowEmptyLambdas = (default)false
+allowEmptyCatches = (default)false
+allowEmptySwitchBlockStatements = true
+ignoreEnhancedForColon = (default)true
+tokens = (default)ASSIGN, BAND, BAND_ASSIGN, BOR, BOR_ASSIGN, BSR, BSR_ASSIGN, BXOR, \
+         BXOR_ASSIGN, COLON, DIV, DIV_ASSIGN, DO_WHILE, EQUAL, GE, GT, LAMBDA, LAND, \
+         LCURLY, LE, LITERAL_CATCH, LITERAL_DO, LITERAL_ELSE, LITERAL_FINALLY, \
+         LITERAL_FOR, LITERAL_IF, LITERAL_RETURN, LITERAL_SWITCH, LITERAL_SYNCHRONIZED, \
+         LITERAL_TRY, LITERAL_WHILE, LOR, LT, MINUS, MINUS_ASSIGN, MOD, MOD_ASSIGN, \
+         NOT_EQUAL, PLUS, PLUS_ASSIGN, QUESTION, RCURLY, SL, SLIST, SL_ASSIGN, SR, \
+         SR_ASSIGN, STAR, STAR_ASSIGN, LITERAL_ASSERT, TYPE_EXTENSION_AND
+
+
+*/
+
+//non-compiled with javac: Compilable with Java17
+package com.puppycrawl.tools.checkstyle.checks.whitespace.whitespacearound;
+
+public class InputWhitespaceAroundSwitchCasesParensWithAllowEmptySwitchBlockStatements {
+
+    void m(int k) {
+        switch (k) {
+            case 1 :
+                System.out.println("1");
+            case 2: { }
+            case 3:
+            case 4: {}
+            default: {}
+        }
+    }
+
+    void m2(int k) {
+        switch (k) {
+            case 1 -> System.out.println("1");
+            case 2 -> { }
+            case 3 -> {}
+            default -> {}
+        }
+    }
+
+    void m3(int k) {
+        switch (k) {
+            case 1 -> {System.out.println("1");}
+            // 2 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            default -> { }
+        }
+    }
+
+    void m4(int k) {
+        switch (k) {
+            case 1 : {System.out.println("1");}
+            // 2 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            default : { }
+        }
+    }
+
+    void m5(int k) {
+        {}
+        // 2 violations above:
+        //                  ''{' is not followed by whitespace.'
+        //                  ''}' is not preceded with whitespace.'
+
+        {System.out.println("1");}
+        // 2 violations above:
+        //                  ''{' is not followed by whitespace.'
+        //                  ''}' is not preceded with whitespace.'
+
+    }
+
+    void m6(int k) {
+        switch (k) {
+            case 1: {} {}
+            case 2: { } { }
+            case 3: {;}
+            // 2 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            case 4: {{}}
+            // 6 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''{' is not followed by whitespace.'
+            //                  ''{' is not preceded with whitespace.'
+            //                  ''}' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            case 5: { {} }
+            // 2 violations above:
+            //                  ''{' is not followed by whitespace.'
+            //                  ''}' is not preceded with whitespace.'
+            case 6: { { } }
+            case 7:  {
+
+            }break;
+        }
+    }
+}

--- a/src/xdocs/checks/whitespace/whitespacearound.xml
+++ b/src/xdocs/checks/whitespace/whitespacearound.xml
@@ -31,7 +31,8 @@ public @interface Beta {} // empty annotation type
           may optionally be exempted from the policy using the <code>
           allowEmptyMethods</code>, <code>allowEmptyConstructors</code>,
           <code>allowEmptyTypes</code>, <code>allowEmptyLoops</code>,
-          <code>allowEmptyLambdas</code> and <code>allowEmptyCatches</code>
+          <code>allowEmptyLambdas</code>, <code>allowEmptyCatches</code> and
+          <code>allowEmptySwitchBlockStatements</code>
           properties.
         </p>
         <p>This check does not flag as violation double brace initialization like:</p>
@@ -107,6 +108,13 @@ try {
               <td><a href="../../property_types.html#boolean">boolean</a></td>
               <td><code>false</code></td>
               <td>4.0</td>
+            </tr>
+            <tr>
+              <td>allowEmptySwitchBlockStatements</td>
+              <td>Allow empty switch blocks and block statements.</td>
+              <td><a href="../../property_types.html#boolean">boolean</a></td>
+              <td><code>false</code></td>
+              <td>10.19.0</td>
             </tr>
             <tr>
               <td>allowEmptyTypes</td>

--- a/src/xdocs/checks/whitespace/whitespacearound.xml.template
+++ b/src/xdocs/checks/whitespace/whitespacearound.xml.template
@@ -31,7 +31,8 @@ public @interface Beta {} // empty annotation type
           may optionally be exempted from the policy using the <code>
           allowEmptyMethods</code>, <code>allowEmptyConstructors</code>,
           <code>allowEmptyTypes</code>, <code>allowEmptyLoops</code>,
-          <code>allowEmptyLambdas</code> and <code>allowEmptyCatches</code>
+          <code>allowEmptyLambdas</code>, <code>allowEmptyCatches</code> and
+          <code>allowEmptySwitchBlockStatements</code>
           properties.
         </p>
         <p>This check does not flag as violation double brace initialization like:</p>

